### ISSUE TITLE
Potential wrong check - check_token_ratio

### ISSUE THREAD

**Issue by hjchan** [askibin/perpetuals#18]
`check_token_ratio` shouldn't be check inside `add_collateral`, `remove_collateral` and `open_position`. the collateral didn't contribute to the liquidity pool, so it shouldn't be check here.

**Comment by mzs-dev** [comment]
Thats true with the current implementation but speaking of design, however collateral must be considered as part of the pool to facilitate collateral valuation in nominal USD value.

**Comment by askibin** [comment]
when exactly do you think collateral should be considered part of a pool? If it is always the case, then the amount of LP tokens you get can be easily gamed by opening a position with large collateral, depositing liquidity, and closing the position right after.

**Comment by hjchan** [comment]
I think it shouldn't be considered to be part of the pool, the value will be affected when the pool value fluctuates. But there is another problem I notice, the collateral is withdrawn in USD value instead of token value, but we are not collecting stable when it is long position, so there will be different in value when position close.

I can think of 2 proposals here, lets discuss and correct me if I am wrong

Proposal 1:
All positions must be open using stable tokens, no matter long or short. But the locked token is different and it depends on the side and token. This allows us to record all collateral in USD value.

Proposal 2:
Using the same logic now, long with the same token and short with stable, but we calculate the collateral in token value. when the position close, the collateral with profit/loss needs to swap to the user's desired output token 

I will prefer proposal 1 as it makes things easier because everything calculates in USD value and it is less confuse to user on which token need to use to long or short

**Comment by askibin** [comment]
Collateral is always recorded in USD but could be collected in the same token or stable.

By collecting collateral in the same tokens for long positions but recording it in USD, the protocol reduces the amount of profit payoffs when a trader is right. For example, if the position is in SOL and it goes 10x, so does the collateral. But since it was recorded in USD, only 1/10 will be returned back to the trader.

If, on the contrary, SOL goes down 50% and the trader who opened a long position is actually losing the money, then the protocol will profit less because it will need to return 2x collateral (but it will still profit).

Same story with short positions. By collecting the collateral in stablecoins, the protocol loses less when the trader is right and profits less when they are wrong.

Protocol also records the amount that was collected. So the trader can't withdraw more collateral (with remove_collateral instruction) than was deposited. Also, there is a check that the total profit payoff doesn't exceed the locked amount + collateral amount (in tokens, not USD).

**Comment by hjchan** [comment]
Thanks for the explanation, didn't know this is to hedge part of the risk.

With that being said, the collateral should actually be contributed to the pool and it should be recorded in USD value instead of token value. There are a few changes we might need to change in my opinion, please correct me if I am wrong

1. the collateral should be contributed to the pool and added to asset.owned, but it should also added into asset.collateral, then when we calculate `get_assets_under_management_usd`, it should be `assset.owned - asset.collateral`
2. update asset.collateral to USD amount instead of the token amount

**Comment by askibin** [comment]
assets, that are part of the pool, are recorded under assets.owned. Lets see how assets.owned are used:
1. As locked liquidity for potential profit payoffs.
2. As amount available to LPs for withdrawal.
3. As amount available for swap.
4. To compute LP tokens to mint/burn when liquidity is added or removed.

None of the above should include the traders' collateral. We could add the collateral to assets.owned and then subtract it upon every caclulation but why?

**Comment by hjchan** [comment]
Because when the collateral is recorded in USD value instead of token amount, the value of the collateral will change over time. 
if a user deposited 100 SOL of collateral at $20. below are a few examples of how pool values should reflect 
1. SOL: $20, collateral: $2000, pool: neutral
2. SOL: $30, collateral: $2000, pool: +$1000 (we need to add the collateral to assets.owned and subtract it to reflect this)
3. SOL: $10, collateral: $2000, pool: -$1000

**Comment by mzs-dev** [comment]
> when exactly do you think collateral should be considered part of a pool? If it is always the case, then the amount of LP tokens you get can be easily gamed by opening a position with large collateral, depositing liquidity, and closing the position right after.

By saying collateral ust be conisdered part of the pool, I don't exactly conisder it to be a profit to the pool, rather the idea is to make add_collateral equivalnet to add_liquidtity **mathematically** where the liquidity added as collateral mints equivalent LP tokens (held in an escrow) to act as the actual collateral , which in no way allows any minuplation on addition/redemption of liquidty to/from the pool. These collateral-LP tokens can be redeemed for underlying liquidity when removing collateral or closing the Position. Again, this method relies on allocating the nominal USD equivalent value of the LP-tokens instead of an absolute usd value for the collateral but it minimises any loss to the pool with regards to compensation on collateral which lost usd_value due to downards price action.

Also, this method primairly focuses on improving capital efficiency of the pool by actively utilising deposited collateral, which introduces a risk where the pool ultimaitely functioning at 100% utlisation would result in insufficent availabile liquidity to reimburse collateral + pnl. Locking up the collateral is one possible solution here to mitigate the risk.

**Comment by askibin** [comment]
first of all, collateral in token amount (not in USD) is tracked already for the custody:
https://github.com/askibin/perpetuals/blob/master/programs/perpetuals/src/state/custody.rs#L67

and the position:
https://github.com/askibin/perpetuals/blob/master/programs/perpetuals/src/state/position.rs#L49

I see the idea of using collateral-LP tokens, but it makes sense only if the collateral amount you are entitled to get back could change over time. If it is fixed, you don't need to do all the math since it is just simply recorded in the position account. 

Users can't remove more collateral than they deposited, this is ensured here:
https://github.com/askibin/perpetuals/blob/master/programs/perpetuals/src/instructions/remove_collateral.rs#L145

And users can't withdraw more PnL than the collateral amount + locked amount, ensured here:
https://github.com/askibin/perpetuals/blob/master/programs/perpetuals/src/state/pool.rs#L157

Let me know if you think checks above are insufficient or using collateral-LP tokens has other advantages.

**Comment by hjchan** [comment]
Users might deposit 1 SOL as collateral but get less SOL when closing position if the SOL price goes up and get more SOL when SOL price go down (technically wouldn't happen because user will lose money here), because it is calculate in USD value. 

We should compare the USD value in both situations instead of the token value.
[master/programs/perpetuals/src/instructions/remove_collateral.rs#L145](https://github.com/askibin/perpetuals/blob/master/programs/perpetuals/src/instructions/remove_collateral.rs?rgh-link-date=2023-03-13T22%3A11%3A02Z#L145)
[master/programs/perpetuals/src/state/pool.rs#L157](https://github.com/askibin/perpetuals/blob/master/programs/perpetuals/src/state/pool.rs?rgh-link-date=2023-03-13T22%3A11%3A02Z#L157)

Besides, after the collateral token is removed from custody, the remaining amount didn't land anywhere and is not counted in the book. I think it should be counted in the book when the user add collateral but we minus out the USD value when counting AUM

**Comment by mzs-dev** [comment]
Exactly, the discrepency in calcualtions needs to be addressed accordingly, hence i proposed the use of collateral-LP tokens to standardise USD value calculations for both pnl and collateral

**Comment by askibin** [comment]
I think we are trying to solve multiple questions at once. 

Lets set aside the question if the collateral is part of the pol or not for now. And answer the question of what should be used as the collateral.

First of all, what is locked for short positions will be changed after we get this merged: https://github.com/askibin/perpetuals/pull/10 Stablecoins will always be locked for the shorts. I'm just holding off on that before fee calculations are updated. Let's pretend we lock the same token when the user opens a long position and USDC when they opens a short. 
I.e. for longs: user deposits SOL as the collateral, we lock SOL
for shorting SOL: user deposits USDC as the collateral, we lock USDC

Now, @hjchan proposal is to use a stablecoin for both cases:
```
Proposal 1:
All positions must be open using stable tokens, no matter long or short. But the locked token is different and it depends on the side and token. This allows us to record all collateral in USD value.
```
I.e. for longs: user deposits USDC as the collateral, we lock SOL
for shorting SOL: user deposits USDC as the collateral, we lock USDC

As I stated before, there is a reason for receiving the collateral in the same token for longs:
```
By collecting collateral in the same tokens for long positions but recording it in USD, the protocol reduces the amount of profit payoffs when a trader is right. For example, if the position is in SOL and it goes 10x, so does the collateral. But since it was recorded in USD, only 1/10 will be returned back to the trader.
```
Basically, protocol reduces risks when traders are right. And yes, it is symmetrical, i.e. protocol loses less when the trader is right but also profits less when the trader is wrong. If when the price goes down, we still count the collateral as 1 SOL and do not use USD value, then the trader would encounter larger losses and feel that the system is unfair.

Anyway, how the collateral is recorded and used in PnL calcs can be adjusted. Let's agree first on what tokens are used to collect the collateral.

**Comment by hjchan** [comment]
Definitely agree with you on this issue. we lock the same token when the users open a long position and USDC when they open a short. And record all the user collateral in USD value

**Comment by askibin** [comment]
great, now lets tackle this:
> Users might deposit 1 SOL as collateral but get less SOL when closing position if the SOL price goes up and get more SOL when SOL price go down (technically wouldn't happen because user will lose money here), because it is calculate in USD value.

Do we all agree that this is the right approach? I.e. it should be symmetrical, since protocol is losing less when the SOL price goes up, it should also win less when the price goes down? This is because the collateral is always counted in USD. 

So in the example when SOL price goes down from $100 to $50 and the trader has a long position with 1 SOL collateral and no leverage (==1), the trader will be able to close the position receiving exactly the same amount, i.e. 1 SOL minus fees minus spreads minus borrow rate. This is because Payoff w/o fees = 100 + (50-100)*1 = $50  (which is 1 SOL)

**Comment by mzs-dev** [comment]
This is the most reasonable take subject to the precondition that we take collateral in and lock the same token when the user opens a long position and stable coins when they opens a short, which shall be refelecting in #10 accordingly.

**Comment by askibin** [comment]
alright, the original description of the ticket:
> check_token_ratio shouldn't be check inside add_collateral, remove_collateral and open_position. the collateral didn't contribute to the liquidity pool, so it shouldn't be check here.
has been taken care of in #47 
I'm closing the ticket and we discuss if any other changes are needed in #10
